providers/ldap/ldap_client.go

```diff
@@ -0,0 +1,649 @@
+	Type: "token",
+}}
+
+func (l *LClient) InitializeSearchConfig() *SearchConfig {
+	c := l.ConstantsConfig
+	return &SearchConfig{
+		Server:       l.Config.Server,
+		Port:         l.Config.Port,
+		BindDN:       l.Config.ServiceAccountUsername,
+		BindPassword: l.Config.ServiceAccountPassword,
+		UserSearchAttributes: []string{c.MemberOfAttribute,
+			c.ObjectClassAttribute,
+			l.Config.UserObjectClass,
+			l.Config.UserLoginField,
+			l.Config.UserNameField,
+			l.Config.UserSearchField,
+			l.Config.UserEnabledAttribute},
+		GroupSeachAttributes: []string{c.MemberOfAttribute,
+			c.ObjectClassAttribute,
+			l.Config.GroupObjectClass,
+			l.Config.UserLoginField,
+			l.Config.GroupNameField,
+			l.Config.GroupSearchField},
+	}
+}
+
+func (l *LClient) newConn() (*ldap.Conn, error) {
+	log.Debug("Now creating Ldap connection")
+	searchConfig := l.SearchConfig
+	lConn, err := ldap.Dial("tcp", fmt.Sprintf("%s:%d", searchConfig.Server, searchConfig.Port))
+	if err != nil {
+		return nil, fmt.Errorf("Error %v creating connection", err)
+	}
+	lConn.SetTimeout(time.Duration(l.Config.ConnectionTimeout) * time.Second)
+
+	err = lConn.Bind(l.Config.ServiceAccountUsername, l.Config.ServiceAccountPassword)
+	if err != nil {
+		return nil, fmt.Errorf("Error %v in ldap bind", err)
+	}
+
+	return lConn, nil
+}
+
+// GenerateToken generates token
+func (l *LClient) GenerateToken(jsonInput map[string]string) (model.Token, error) {
+	log.Info("Now generating Ldap token")
+	searchConfig := l.SearchConfig
+
+	//getLdapToken:ADTokenCreator
+	//getIdentities: ADIdentityProvider
+
+	split := strings.Split(jsonInput["code"], ":")
+	username, password := split[0], split[1]
+	externalId := username
+	if l.Config.LoginDomain != "" {
+		externalId = l.Config.LoginDomain + "\\" + username
+	}
+
+	if password == "" {
+		return nilToken, fmt.Errorf("Failed to login, password not provided")
+	}
+
+	lConn, err := l.newConn()
+	if err != nil {
+		return nilToken, fmt.Errorf("Error %v creating connection", err)
+	}
+	log.Info("Binding username password")
+	err = lConn.Bind(externalId, password)
+	if err != nil {
+		return nilToken, fmt.Errorf("Error %v in ldap bind", err)
+	}
+	defer lConn.Close()
+	samName := username
+	if strings.Contains(username, "\\") {
+		samName = strings.SplitN(username, "\\\\", 2)[1]
+	}
+	query := "(" + l.Config.UserLoginField + "=" + samName + ")"
+	if l.AccessMode == "required" {
+		groupFilter, err := l.getAllowedIdentitiesFilter()
+		if err != nil {
+			return nilToken, err
+		}
+		if len(groupFilter) > 1 {
+			groupQuery := "(&" + query + groupFilter + ")"
+			query = groupQuery
+		}
+
+	}
+
+	search := ldap.NewSearchRequest(l.Config.Domain,
+		ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
+		query,
+		searchConfig.UserSearchAttributes, nil)
+
+	return l.userRecord(search, lConn)
+}
+
+func (l *LClient) getIdentitiesFromSearchResult(result *ldap.SearchResult) ([]client.Identity, error) {
```

Comment:
We need to put all the log statements added in TRACE in Java code in right places. They are usually needed for debugging customer issues.